clang/lib/DPCT/ASTTraversal.cpp

```diff
@@ -13945,6 +13945,90 @@ void FFTFunctionCallRule::runRule(const MatchFinder::MatchResult &Result) {
+  };
+  auto virtualmemoryEnum = [&]() {
+    return hasAnyName("CU_MEM_ALLOCATION_TYPE_PINNED",
+                      "CU_MEM_LOCATION_TYPE_DEVICE",
+                      "CU_MEM_ACCESS_FLAGS_PROT_READWRITE",
+                      "CU_MEM_ALLOC_GRANULARITY_RECOMMENDED");
+  };
+  MF.addMatcher(
+      callExpr(callee(functionDecl(virtualmemoryAPI()))).bind("vmCall"), this);
+  MF.addMatcher(
+      typeLoc(loc(qualType(hasDeclaration(namedDecl(virtualmemoryType())))))
+          .bind("vmType"),
+      this);
+  MF.addMatcher(
+      declRefExpr(to(enumConstantDecl(virtualmemoryEnum()))).bind("vmEnum"),
+      this);
+}
+
+void VirtualMemoryRule::runRule(
+    const ast_matchers::MatchFinder::MatchResult &Result) {
+  auto &SM = DpctGlobalInfo::getSourceManager();
+  if (const CallExpr *CE = getNodeAsType<CallExpr>(Result, "vmCall")) {
+    ExprAnalysis EA(CE);
+    emplaceTransformation(EA.getReplacement());
+    EA.applyAllSubExprRepl();
+  }
+  if (auto TL = getNodeAsType<TypeLoc>(Result, "vmType")) {
+    auto TypeStr =
+        DpctGlobalInfo::getTypeName(TL->getType().getUnqualifiedType());
+    if (!DpctGlobalInfo::useExpVirtualMemory()) {
+      report(TL->getBeginLoc(), Diagnostics::TRY_EXPERIMENTAL_FEATURE, false,
+             TypeStr, "--use-experimental-features=virtual_memory");
+      return;
+    }
+    if (!DpctGlobalInfo::isInAnalysisScope(
+            SM.getSpellingLoc(TL->getBeginLoc()))) {
+      return;
+    }
+    auto Range = getDefinitionRange(TL->getBeginLoc(), TL->getEndLoc());
+    auto BeginLoc = Range.getBegin();
+    auto EndLoc = Range.getEnd();
+
+    if (SM.isWrittenInScratchSpace(SM.getSpellingLoc(TL->getBeginLoc()))) {
+      BeginLoc = SM.getExpansionRange(TL->getBeginLoc()).getBegin();
+      EndLoc = SM.getExpansionRange(TL->getBeginLoc()).getEnd();
+    }
+    std::string Str =
+        MapNames::findReplacedName(MapNames::TypeNamesMap, TypeStr);
+    if (!Str.empty()) {
+      auto Len = Lexer::MeasureTokenLength(
+          EndLoc, SM, DpctGlobalInfo::getContext().getLangOpts());
+      Len += SM.getDecomposedLoc(EndLoc).second -
+             SM.getDecomposedLoc(BeginLoc).second;
+      emplaceTransformation(new ReplaceText(BeginLoc, Len, std::move(Str)));
+      return;
+    }
+  }
+  if (auto *E = getNodeAsType<DeclRefExpr>(Result, "vmEnum")) {
+    std::string EnumName = E->getNameInfo().getName().getAsString();
+    if (!DpctGlobalInfo::useExpVirtualMemory()) {
+      report(E->getBeginLoc(), Diagnostics::TRY_EXPERIMENTAL_FEATURE, false,
+             EnumName, "--use-experimental-features=virtual_memory");
+      return;
+    }
+    auto Search = EnumConstantRule::EnumNamesMap.find(EnumName);
+    if (Search == EnumConstantRule::EnumNamesMap.end()) {
+      report(E->getBeginLoc(), Diagnostics::API_NOT_MIGRATED, false, EnumName);
+      return;
+    }
+    emplaceTransformation(new ReplaceStmt(E, Search->second->NewName));
+  }
+}
+
+REGISTER_RULE(VirtualMemoryRule, PassKind::PK_Migration)
```

Comment:
```suggestion
REGISTER_RULE(VirtualMemRule, PassKind::PK_Migration)
```

Reply:
can we register the rule only when --use-experimental-features=virtual_mem is given?

Reply:
we can't, because dpct need emit warning message to guide user use this extension.